### PR TITLE
8276700: Improve java.lang.ref.Cleaner javadocs

### PR DESCRIPTION
**Pull request by hendriks73**
Trivial improvement.

Explicitly show how to create a `Cleaner` instance using `Cleaner.create()`.
Repeat (again) in the code example that the `State` `Runnable `should be implemented as static class and not reference the instance to be cleaned, to make the point even more clear to those people who never read the javadoc *prose*.

I have signed the OCA a while back as [hschreiber](https://openjdk.java.net/census#hschreiber).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276700](https://bugs.openjdk.java.net/browse/JDK-8276700): Improve java.lang.ref.Cleaner javadocs

### Reviewers
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6076/head:pull/6076` \
`$ git checkout pull/6076`

Update a local copy of the PR: \
`$ git checkout pull/6076` \
`$ git pull https://git.openjdk.java.net/jdk pull/6076/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6076`

View PR using the GUI difftool: \
`$ git pr show -t 6076`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6076.diff">https://git.openjdk.java.net/jdk/pull/6076.diff</a>

</details>

### CODE REVIEW

**Comment by shipilev** [comment]
Submitted the RFE for it. Please change the PR title to "8276700: Improve java.lang.ref.Cleaner javadocs" to let bots hook it up.

File: src/java.base/share/classes/java/lang/ref/Cleaner.java [hunk 1]
```diff
@@ -87,8 +87,10 @@
  * <pre>{@code
  * public class CleaningExample implements AutoCloseable {
  *        // A cleaner, preferably one shared within a library
- *        private static final Cleaner cleaner = <cleaner>;
+ *        private static final Cleaner cleaner = Cleaner.create();
  *
+ *        // Static state class, capturing information necessary for
+ *        // cleanup, but no reference to the instance being cleaned
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/ref/Cleaner.java:93]:
```suggestion
 *        // State class captures information necessary for cleanup.
 *        // It must hold no reference to the instance being cleaned
 *        // and therefore it is a static inner class in this CleaningExample.
```

Would this be clearer?

Comment (mlchung) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:93]: This is what I suggested and makes it clear that *must hold no reference to the instance being cleaned*.  Maybe you didn't notice it's still there? 
 
```
 *        // State class captures information necessary for cleanup.
 *        // It must hold no reference to the instance being cleaned
 *        // and therefore it is a static inner class in this CleaningExample
 ```

File: src/java.base/share/classes/java/lang/ref/Cleaner.java [hunk 2]
```diff
@@ -87,8 +87,10 @@
  * <pre>{@code
  * public class CleaningExample implements AutoCloseable {
  *        // A cleaner, preferably one shared within a library
- *        private static final Cleaner cleaner = <cleaner>;
+ *        private static final Cleaner cleaner = Cleaner.create();
```
[the remark this answers is not in the document]
Comment (hendriks73) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:90]: You have a point.

BUT, at least it's a working example and not some pseudo code. We do want to move to working example code long term, don't we?

When I see <cleaner>, I'm just wondering what those <> type operators are good for here...

[the remark this answers is not in the document]
Comment (RogerRiggs) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:90]: A cleaner can/should be shared within some scope and purpose, in this case the example class .
Each cleaner has a dedicated Thread so its not a lightweight object and should not be proliferated.
(Loom may be able to use Virtual Threads).

The reasons to not share are a Cleaner are based on possible interference between the cleanup callbacks.
If they share a thread and are non-trivial, it might slow other cleaners. If the cleaner is created and shared
for a particular purpose it will share the thread resource and still be efficient.

File: src/java.base/share/classes/java/lang/ref/Cleaner.java (hunk 1 above)
Comment (bchristi-git) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:93]: I like this suggestion - calling out that the static-ness of the State class is due to the requirement to not hold any references.

File: src/java.base/share/classes/java/lang/ref/Cleaner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bchristi-git) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:90]: The new example Cleaner instance _is_ shared, though on a pretty small scale (just among instances of CleaningExample).  A demonstration of larger scale sharing of a Cleaner instance would be out of scope for this example.

[the remark this answers is not in the document]
Comment (hendriks73) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:90]: Let me add, why I have raised this issue.

I was going to migrate some older code which uses the `finalize()` method to the `Cleaner` mechanism. New it it, there seemed to be two pitfalls:

1. Understanding the whole "don't capture an instance reference in your state object"
2. Copying the example (which was in a non-working state, due to pseudo code) and making it work for me

With the improvement suggestions, I was trying help people who *only* read the code sample (many do), to become aware of 1. and help them getting going with 2, simply because it's something they can copy and run.

[the remark this answers is not in the document]
Comment (anthonyvdotbe) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:90]: > When I see `<cleaner>`, I'm just wondering what those <> type operators are good for here...

What about just replacing `<cleaner>` with `...` then? The `State` constructor and its invocation also have an ellipsis.

> BUT, at least it's a working example and not some pseudo code.

The example is still not compilable due to the remaining ellipses.

> We do want to move to working example code long term, don't we?

I agree that examples should be compilable *in the raw Javadoc*. However, in the rendered Javadoc, using ellipses is a well-understood way to keep examples concise and devoid of irrelevant and/or usecase-dependent code. Moreover, when developers copy-paste the example, they'll immediately be pointed to all the places where they need to fill in the blanks, make a choice for a trade-off, etc. On the other hand, by hard-coding a (suboptimal) choice, developers who merely copy-paste the example are unlikely to reconsider the trade-off.

> The new example Cleaner instance _is_ shared, though on a pretty small scale (just among instances of CleaningExample).

True, but my point was that the comment says "shared *within a library*". So to me it's confusing to have a comment saying "it's preferred to do A", and then have the code do B on the next line.

So I propose to either:
* revert the current change & simply replace `<cleaner>` with `...`
* update the comment to say: "A cleaner (preferably one shared within a library, but for the sake of example, a new one is created here)"

Actually, to have the line be compilable, and at the same time (attempt to) force developers to consider the trade-off, it could be changed to something like:

```java
private static final Cleaner cleaner = createCleaner();

private static Cleaner createCleaner() {
    // A cleaner, preferably one shared within a library
    throw new UnsupportedOperationException("TBD");
}
```

**Comment by RogerRiggs** [comment]
As for compilable code, this may be a good place to use Snippets.  
[JEP 413: Code Snippets in Java API Documentation](https://openjdk.java.net/jeps/413)
They are new but may offer some value add here.

Good standalone examples aren't always good programming practice.

File: src/java.base/share/classes/java/lang/ref/Cleaner.java (hunk 1 above)
Comment (hendriks73) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:93]: > Maybe you didn't notice it's still there?

Yeah. Sorry, @mlchung. Good suggestion!

File: src/java.base/share/classes/java/lang/ref/Cleaner.java [hunk 3]
```diff
@@ -87,8 +87,12 @@
...
  *
+ *        // Static state class, capturing information necessary for
+ *        // State class captures information necessary for cleanup.
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/ref/Cleaner.java:93]: I think it would be simpler to just drop L92, meaning start with "State class captures ..." rather than having "state class" and "State class" in the same sentence.

File: src/java.base/share/classes/java/lang/ref/Cleaner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hendriks73) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:90]: This is getting too complicated...

It's a code *example* with a very clear comment that explains a best practice:
```java
// A cleaner, preferably one shared within a library
private static final Cleaner cleaner = Cleaner.create();
```

We cannot really show the best practice in this example without making the example itself more complicated. IMHO, introducing an extra factory method here adds nothing but complexity and makes the example more difficult to understand (that aside, it should probably be something like `MyLibrary.getCleaner()` and not a `createXyz()` method).

I still much more prefer `cleaner = Cleaner.create();` over `cleaner = <cleaner>` (which really is no better in any way, shape or form and creates more questions than it provides answers) or `cleaner = ...`, which again does not answer the question of how to get a cleaner instance—something I asked myself when trying to use the API. In fact *how to get a cleaner instance* is not explained by the current javadocs *at all* and it's something one *must* know to use this API.

Here's our chance to show how it *can* be done, even if it's not ideal, because it does not demonstrate the one-cleaner per library recommendation (only mentions it).

And yes, those ellipsis in the `State` class code—I'd love for them to be gone, too. It would make the example less abstract and easier to understand (what's this state anyway? in most cases it's probably a reference to a state, e.g. a native pointer, i.e. a `long`). But, admittedly, that's difficult.

So, to summarize, there is always a tradeoff between making an example easy to understand, not too complex, but still conveying the most important information and best practices. In the end it's a matter of opinion. In this case, I will stick to my original code change suggestion, because it adds value (answers the question of how to get/create a cleaner instance).

File: src/java.base/share/classes/java/lang/ref/Cleaner.java (hunk 3 above)
Comment (hendriks73) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:92]: I have removed that line and also changed `in this CleaningExample` -> `in this example`, because it is obvious.

File: src/java.base/share/classes/java/lang/ref/Cleaner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (anthonyvdotbe) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:90]: In short: the current code

```java
// A cleaner, preferably one shared within a library
private static final Cleaner cleaner = <cleaner>;
```

is unhelpful to you, and the proposed code

```java
// A cleaner, preferably one shared within a library
private static final Cleaner cleaner = Cleaner.create();
```

is confusing to me.

Trying to find a compromise, I'm merely asking that the comment be clarified:

```java
// A cleaner (preferably one shared within a library, but for the sake of example, a new one is created here)
private static final Cleaner cleaner = Cleaner.create();
```

What do you think?

[the remark this answers is not in the document]
Comment (hendriks73) [reply src/java.base/share/classes/java/lang/ref/Cleaner.java:90]: Sounds great!
I'll change the code.
Thanks for suggesting this compromise.

**Review by RogerRiggs — APPROVED** [review]
LGTM

**Review by mlchung — APPROVED** [review]

**Comment by hendriks73** [comment]
@RogerRiggs @mlchung Could either of you please sponsor this? Thanks so much and hopefully a great start into 2022!
